Sources/SwiftGRPCNIO/HTTP1ToRawGRPCServerCodec.swift

```diff
@@ -27,111 +27,200 @@ public enum RawGRPCServerResponsePart {
 ///
 /// The translation from HTTP2 to HTTP1 is done by `HTTP2ToHTTP1ServerCodec`.
 public final class HTTP1ToRawGRPCServerCodec {
-  private enum State {
+  internal var inboundState = InboundState.expectingHeaders
+  internal var outboundState = OutboundState.expectingHeaders
+
+  private var buffer: NIO.ByteBuffer?
+
+  // 1-byte for compression flag, 4-bytes for message length.
+  private let protobufMetadataSize = 5
+}
+
+extension HTTP1ToRawGRPCServerCodec {
+  enum InboundState {
     case expectingHeaders
-    case expectingCompressedFlag
-    case expectingMessageLength
-    case receivedMessageLength(UInt32)
-
-    var expectingBody: Bool {
-      switch self {
-      case .expectingHeaders: return false
-      case .expectingCompressedFlag, .expectingMessageLength, .receivedMessageLength: return true
-      }
+    case expectingBody(Body)
+    // ignore any additional messages; e.g. we've seen .end or we've sent an error and are waiting for the stream to close.
+    case ignore
+
+    enum Body {
+      case expectingCompressedFlag
+      case expectingMessageLength
+      case receivedMessageLength(UInt32)
     }
   }
 
-  private var state = State.expectingHeaders
+  enum OutboundState {
+    case expectingHeaders
+    case expectingBodyOrStatus
+    case ignore
+  }
+}
 
-  private var buffer: NIO.ByteBuffer?
+extension HTTP1ToRawGRPCServerCodec {
+  struct StateMachineError: Error, GRPCStatusTransformable {
```

Comment:
Make this `struct` and `message` public (and/or `CustomStringConvertible`), so that library users can inspect the error?